File: app/src/main/java/ru/laushkina/rates/repository/RatesInMemoryCache.java
package ru.laushkina.rates.repository;

import androidx.annotation.NonNull;

import java.util.List;
import java.util.Map;

import io.reactivex.subjects.PublishSubject;
import io.reactivex.subjects.Subject;
import ru.laushkina.rates.model.Rate;

public class RatesInMemoryCache {
    private static final Object lock = new Object();
    private static volatile RatesInMemoryCache INSTANCE;

    @NonNull
    private Subject<List<Rate>> observable = PublishSubject.create();

    public static RatesInMemoryCache getInstance() {
        if (INSTANCE == null) {
            synchronized (lock) {
                if (INSTANCE == null) {
                    INSTANCE = new RatesInMemoryCache();
                }
            }
        }
        return INSTANCE;
    }

    @NonNull
    public Subject<List<Rate>> getCachedRates() {
        return observable;
    }

    public void setCachedRates(@NonNull List<Rate> rates) {
        observable.onNext(rates);
    }
}

File: app/src/main/java/ru/laushkina/rates/repository/RatesInMemoryCache.kt
package ru.laushkina.rates.repository

import io.reactivex.subjects.PublishSubject
import io.reactivex.subjects.Subject
import ru.laushkina.rates.model.Rate

class RatesInMemoryCache {
    private val observable: Subject<List<Rate>> = PublishSubject.create()

    companion object {
        private val lock = Any()

        @Volatile
        private var INSTANCE: RatesInMemoryCache? = null

        // TODO
        fun getInstance(): RatesInMemoryCache {
            if (INSTANCE == null) {
                synchronized(lock) {
                    if (INSTANCE == null) {
                        INSTANCE = RatesInMemoryCache()
                    }
                }
            }
            return INSTANCE!!
        }
    }

    fun getCachedRates(): Subject<List<Rate>> {
        return observable
    }

    fun setCachedRates(rates: List<Rate>?) {
        if (rates != null) {
            observable.onNext(rates)
        }
    }
}